### PR TITLE
Convert JSONB data to text when insensitive case sort is requested

### LINKED ISSUE #146: jsonb sort: convert to text
When sorting on a jsonb sub-field, the function `jsonb_extract_path` is used.
This function returns a `jsonb` result type. When sorting 'ic' (case insensitive), `lower` function is used. But `lower` function expects 'text' as argument and does not work on jsonb.
The returned error is:  `No function matches the given name and argument types. You might need to add explicit type casts.`

My proposal is: when you know an 'ic' comparison is expected, you could use `jsonb_extract_path_text` function instead of `jsonb_extract_path`, then lower function works.

**How to fix it in SortUtils.java:**

Add boolean parameter `caseInsensitive` to method `sortExpressionOfJson`.
Use boolean to return `jsonb_extract_path_text` function:
```
if (caseInsensitive) {
  return builder.function("jsonb_extract_path_text", String.class, args.toArray(Expression[]::new));
} else {
  return builder.function("jsonb_extract_path", String.class, args.toArray(Expression[]::new));
}
```

In method `sortToJpaOrder`, before calling `sortExpressionOfJson`, find out if 'ic' is expected: 
`final boolean caseInsensitive = parts.length > 2 && "ic".equalsIgnoreCase(parts[2]);`

### PR DESCRIPTION
**Pull request by itineric**
close #146

### CODE REVIEW

**Comment by perplexhub** [comment]
@itineric Thanks for your contribution.
@ng-galien Could you please help to review? Thanks.

**Comment by ng-galien** [comment]
> @itineric Thanks for your contribution. @ng-galien Could you please help to review? Thanks.

Hi  @perplexhub, I check it.

File: rsql-jpa/src/main/java/io/github/perplexhub/rsql/SortUtils.java [hunk 1]
```diff
@@ -64,16 +64,16 @@ private static Order sortToJpaOrder(final String[] parts, final SortSupport sort
                         sortSupport.getJoinHints(), sortSupport.getProcedureWhiteList(),
                         sortSupport.getProcedureBlackList());
 
+        final boolean ic = parts.length > 2 && "ic".equalsIgnoreCase(parts[2]);
```
Comment (ng-galien) [inline rsql-jpa/src/main/java/io/github/perplexhub/rsql/SortUtils.java:67]: We probably don't need to add an extra parameter for changing jsonb function.
Casting jsonb extract path to text in JPA should be sufficient, and we can remove unchecked cast.
`@SuppressWarnings("unchecked")`

File: rsql-jpa/src/main/java/io/github/perplexhub/rsql/SortUtils.java [hunk 2]
```diff
@@ -64,16 +64,16 @@ private static Order sortToJpaOrder(final String[] parts, final SortSupport sort
...
             final boolean isJson = JsonbSupport.isJsonType(rsqljpaContext.getAttribute());
             return isJson
-                    ? sortExpressionOfJson(rsqljpaContext, string, sortSupport.getPropertyPathMapper(), builder)
+                    ? sortExpressionOfJson(rsqljpaContext, string, sortSupport.getPropertyPathMapper(), builder, ic)
                     : rsqljpaContext.getPath();
         });
 
-        if (parts.length > 2 && "ic".equalsIgnoreCase(parts[2])
-            && String.class.isAssignableFrom(propertyExpression.getJavaType())) {
+        if (ic && String.class.isAssignableFrom(propertyExpression.getJavaType())) {
```
Comment (ng-galien) [inline rsql-jpa/src/main/java/io/github/perplexhub/rsql/SortUtils.java:76]:
```java
if (parts.length > 2 && "ic".equalsIgnoreCase(parts[2])
    && String.class.isAssignableFrom(propertyExpression.getJavaType())) {
    //Cast to text
    propertyExpression = cb.lower(propertyExpression.as(String.class));
}
```

File: rsql-jpa/src/main/java/io/github/perplexhub/rsql/SortUtils.java [hunk 3]
```diff
@@ -88,12 +88,14 @@ private static Order sortToJpaOrder(final String[] parts, final SortSupport sort
...
      * @param builder  the criteria builder
+     * @param ic a flag indicating if sort is case insensitive
      * @return the jsonb expression
      */
     private static Expression<?> sortExpressionOfJson(RSQLJPAContext context,
                                                       String property,
                                                       Map<String, String> mapping,
-                                                      CriteriaBuilder builder) {
+                                                      CriteriaBuilder builder,
+                                                      boolean ic) {
```
Comment (ng-galien) [inline rsql-jpa/src/main/java/io/github/perplexhub/rsql/SortUtils.java:98]: we don't need this extra parameter has the Expression will be casted to String later.

File: rsql-jpa/src/test/java/io/github/perplexhub/rsql/RSQLJPASupportPostgresJsonTest.java [hunk 4]
```diff
@@ -602,6 +603,24 @@ private static Stream<Arguments> doesNotConvertInteger() {
         ).filter(Objects::nonNull);
     }
 
+    private static Stream<Arguments> sortByText() {
+        var e1 = new PostgresJsonEntity(Map.of("a", "abc", "b", 1));
```
Comment (ng-galien) [inline rsql-jpa/src/test/java/io/github/perplexhub/rsql/RSQLJPASupportPostgresJsonTest.java:607]: Try to mix string, numbers, booleans and null in a third key to ensure we never raise an exception.

**Review by ng-galien — CHANGES_REQUESTED** [review]
Good point and we have regression test.
Have a look on my comments.

File: rsql-jpa/src/test/java/io/github/perplexhub/rsql/RSQLJPASupportPostgresJsonTest.java (hunk 4 above)
Comment (itineric) [reply rsql-jpa/src/test/java/io/github/perplexhub/rsql/RSQLJPASupportPostgresJsonTest.java:607]: I dont get that comment, you're expecting a "c" key to mix data types? But thats not related to "sorting by text". Maybe  we should add another test with mixed types?

Comment (ng-galien) [reply rsql-jpa/src/test/java/io/github/perplexhub/rsql/RSQLJPASupportPostgresJsonTest.java:607]: Something simple as

```
var e2 = new PostgresJsonEntity(Map.of("a", "ABC", "b", 2, "c", 1));
var e3 = new PostgresJsonEntity(Map.of("a", "DEF", "b", 3, "c", "a"));
var e4 = new PostgresJsonEntity(Map.of("a", "GHI", "b", 4, "c", true));
var e5 = new PostgresJsonEntity(Map.of("a", "ghi", "b", 5, "c", null));
var e6 = new PostgresJsonEntity(Map.of("a", "klm", "b", 6));
```

In fact, we don't care about the result of sorting with `c`, it's just to ensure it does not throw any exception.

Comment (itineric) [reply rsql-jpa/src/test/java/io/github/perplexhub/rsql/RSQLJPASupportPostgresJsonTest.java:607]: Là c’est sûr qu'on se prend une NPE sur cette ligne :
`var e5 = new PostgresJsonEntity(Map.of("a", "ghi", "b", 5, "c", null));`

Mais ok, je vois l’idée, je ferai ça lundi

**Review by ng-galien — APPROVED** [review]

**Comment by ng-galien** [comment]
@itineric nickel, merci!

**Pull request merged**